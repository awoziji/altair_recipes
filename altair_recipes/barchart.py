"""Generate barcharts."""
from .signatures import bivariate_recipe, color
import altair as alt
from autosig import autosig, Signature


@autosig(bivariate_recipe + Signature(color=color(default=None, position=3)))
def barchart(data, x=0, y=1, color=None, height=300, width=400):
    """Generate a barchart."""
    chart = alt.Chart(
        data,
        height=height,
        width=width / ((len(data[x].unique()) if color is not None else 1)),
    ).mark_bar()
    if color is None:
        return chart.encode(x=x, y=y)
    else:
        return chart.encode(x=alt.X(color, axis=None), y=y, color=color, column=x)
