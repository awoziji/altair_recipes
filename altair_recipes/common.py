import altair as alt
from autosig import signature, Signature, param
import pandas as pd


@signature
class Recipe(Signature):
    data = param()
    mark = param(default={}, convert=dict)
    encoding = param(default={}, convert=dict)
    properties = param(default={}, convert=dict)


@signature
class UnivariateRecipe(Recipe):
    column = param(default=0)


@signature
class BivariateRecipe(Recipe):
    x = param(default="x")
    y = param(default="y")


@signature
class MultivariateRecipe(Recipe):
    columns = param(default=None)
    group_by = param(default=None)

    def validate(self):
        if self.columns is None: columns = list(self.data.columns)


def viz_reg_test(test_f):
    def fun(regtest):
        plot = test_f()
        if regtest is not None:
            regtest.write(plot.to_json())
        plot.save(
            test_f.__code__.co_filename + "_" + test_f.__qualname__ + ".html")
        return plot

    return fun


def to_dataframe(data):
    #this is an undocumented hack
    return alt.Chart(data).data


def default(*args):
    return next(x for x in args if x is not None)


# unused should kill?
# def update(d1, d2):
#     d1.update(d2)
#     return d1


def gather(data, key, value, columns):
    return pd.melt(
        data,
        id_vars=[col for col in data.columns if col not in columns],
        value_vars=columns,
        var_name=key,
        value_name=value)
