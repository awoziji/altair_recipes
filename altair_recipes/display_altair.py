from IPython.core.display import display, HTML


def display_html(x):
    return display(HTML(x))


display_html("""
<script src="https://cdn.jsdelivr.net/npm/vega@3"></script>
<script src="https://cdn.jsdelivr.net/npm/vega-lite@2"></script>
<script src="https://cdn.jsdelivr.net/npm/vega-embed@3"></script>
""")


def display_altair_chart(x):
    display_html("""
 <div id="vis"></div>
  <script type="text/javascript">
    var spec = """ + x.to_json() + """;
    var opt = {"renderer": "canvas", "actions": false};
    vegaEmbed("#vis", spec, opt);
  </script>
  """)
