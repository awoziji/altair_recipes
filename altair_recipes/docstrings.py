"""Reusable docstrings elements."""
from enum import Enum


class Docstring(Enum):
    """Reusable docstring elements collected in an enum."""

    data = """data: Altair Data or pandas DataFrame or csv or json file URL
    The data from which the statistical graphics is being generated"""
    column = """column: str or other column selector
    The column containing the data to be used in the graphics"""
    x = """x: str or other columns selector
    The column containing the data associated with the horizontal dimension"""
    y = """y: str or other column selector
    The column containing the data associated with the vertical dimension"""
    columns = """columns: str or list thereof or other column selector
    The column or columns to be used in the graphics"""
    group_by = """group_by: str or other column selector
    The column to be used to group the data when in long form. When group_by is
    specified columns should point to a single column"""
    return_object = """altair.Chart_area
    The graphics object"""
    mark = """mark: dictionary
    Additional arguments to pass to the mark method of Chart"""
    encoding = """encoding: dictionary
    Additional arguments to the encode method of Chart"""
    properties = """properties: dictionary
    Additional arguments to the properties method of Chart"""
    returns = """altair.Chart
    The chart described in the summary"""


def make_docstring(summary, params, returns=Docstring.returns):
    """Make docstrings from simple reusable parts.

    Parameters
    ----------
    summary : str
        The docstring summary.
    params : str
        A list of parameter docstrings. If present in the .docstrings.Docstring
        enum, will be replaced by corresponding value (default meaning for that
        argument)
    returns : str
        The docstring return section. If just "returns", will be replaced by
        .docstrings.Docstring.returns

    Returns
    -------
    str
        A docstring. Assign to __doc__ of desired object.

    """
    params = list(
        map(lambda x: x.value if type(x) == Docstring else x, params))
    returns = returns.value if type(returns) == Docstring else returns
    return "    \n".join([summary, "\nParameters\n---------"] + params +
                         ["\nReturns\n-------", returns])
