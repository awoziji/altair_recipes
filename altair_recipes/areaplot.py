"""areaplots."""
from common import warn_not_distinct
from .signatures import bivariate_recipe, color
import altair as alt
from autosig import autosig, Signature, param
from logging import warning
from enum import Enum


StackType = Enum("StackType", "true false  normalize")


@autosig(
    bivariate_recipe
    + Signature(
        color(default=2, position=3),
        stack=param(
            default=None,
            position=4,
            converter=lambda x: StackType(x) if x is not None else x,
        ),
    )
)
def areaplot(data=None, x=0, y=1, color=2, stack=None, height=600, width=800):
    """Generate a areaplot."""
    warn_not_distinct(data[x])
    if len(set(data[x])) != len(data[x]):
        warning("The relation to plot is not a function")
    if stack is not None:
        x = alt.X(x=x, stack=stack.name)
    return alt.Chart(data=data, height=height, width=width).mark_area().encode(x=x, y=y)
