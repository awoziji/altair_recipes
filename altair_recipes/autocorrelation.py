from .common import to_dataframe, UnivariateRecipe
import altair as alt
from autosig import autosig, signature, param
import numpy as np
import pandas as pd


@signature
class Autocorrelation(UnivariateRecipe):
    max_lag = param(default=None)


@autosig(Autocorrelation)
def autocorrelation(data,
                    column=0,
                    max_lag=None,
                    mark={},
                    encoding={},
                    properties={}):
    data = to_dataframe(data)
    max_lag = data.shape[0] - 1 if max_lag is None else int(max_lag)
    if type(column) is int:
        column = data.columns[0]
    lags = np.arange(0, max_lag + 1)
    _data = pd.DataFrame(
        dict(
            Lag=lags,
            Autocorrelation=[data[column].autocorr(lag=lag) for lag in lags]))
    return alt.Chart(_data).mark_bar(**mark).encode(
        x="Lag:O", y="Autocorrelation" + ":Q",
        **encoding).properties(**properties)
