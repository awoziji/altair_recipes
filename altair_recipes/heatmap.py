"""Heatmap implementation."""
import altair as alt
from .docstrings import make_docstring
from .common import to_dataframe
from math import sqrt


def heatmap_preprocess(data, x, y):
    data = to_dataframe(data)
    nx = sqrt(data.shape[0])
    ny = nx
    return data, nx, ny


def heatmap(
        data,
        x,
        y,
        color,
):
    """See below."""
    data, nx, ny = heatmap_preprocess(data, x, y)
    return alt.Chart(data).mark_rect().encode(
        x=alt.X(x + ':Q', bin=alt.Bin(maxbins=nx)),
        y=alt.Y(y + ':Q', bin=alt.Bin(maxbins=ny)),
        color=alt.Color(
            'average(' + color + '):Q', scale=alt.Scale(scheme='greenblue')))


heatmap.__doc__ = make_docstring(
    heatmap,
    """Generate a heatmap.""",
    additional_params=dict(color="""color: `str` or other column selector
    The color to be used at each `(x,y)` location"""))


def binned_heatmap(
        data,
        x,
        y,
):
    """See below."""
    data, nx, ny = heatmap_preprocess(data, x, y)
    return alt.Chart(data).mark_rect().encode(
        x=alt.X(x + ':Q', bin=alt.Bin(maxbins=nx)),
        y=alt.Y(y + ':Q', bin=alt.Bin(maxbins=ny)),
        color=alt.Color(
            'count(' + x + '):Q', scale=alt.Scale(scheme='greenblue')))


heatmap.__doc__ = make_docstring(
    binned_heatmap,
    summary="Create a heatmap of the count of points in each square")
