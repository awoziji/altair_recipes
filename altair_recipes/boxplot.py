"""Boxplot implementation"""
import altair as alt
from altair_recipes.common import default, gather, to_dataframe


#TODO: inject addional option into graph
#TODO: unclear how to do it with complex graphics such as this, if there is a way
def boxplot(data, columns, group_by=None, mark={}, encoding={}, properties={}):
    data = to_dataframe(data)
    assert type(columns) == str or len(columns) == 1 or group_by is None
    # we are doing wide or long format but not both
    if group_by is None:  #convert wide to long
        x = default(data.columns.name, "variable")
        y = "value"
        data = gather(data, key=x, value=y, columns=columns)
    else:
        x = group_by
        y = columns if type(columns) is str else columns[0]
    #long form assumed from here
    chart = alt.Chart(data)
    chart_bar = chart.mark_bar(filled=False)
    chart_tick = chart.mark_tick()
    min_y = "min(" + y + ")"
    max_y = "max(" + y + ")"
    median_y = "median(" + y + ")"
    q1_bar = chart_bar.encode(x=x, y="q1(" + y + ")", y2=median_y)
    q3_bar = chart_bar.encode(x=x, y=median_y, y2="q3(" + y + ")")

    min_tick = chart_tick.encode(x=x, y=min_y)
    max_tick = chart_tick.encode(x=x, y=max_y)
    rule = chart.mark_rule().encode(
        x=x, y=alt.Y(min_y, axis=alt.Axis(title=y)), y2=max_y)
    return q1_bar + q3_bar + min_tick + max_tick + rule
