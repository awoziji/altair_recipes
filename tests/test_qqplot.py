import numpy as np
import pandas as pd
import altair_recipes as vs

df = pd.DataFrame({
    'Trial A': np.random.normal(0, 0.8, 1000),
    'Trial B': np.random.normal(-2, 1, 1000),
    'Trial C': np.random.uniform(3, 2, 1000)
})

vs.qqplot(df, x='Trial A', y='Trial C')
