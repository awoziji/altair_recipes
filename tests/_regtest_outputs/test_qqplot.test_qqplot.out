{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.4.3.json",
  "config": {
    "view": {
      "height": 300,
      "width": 400
    }
  },
  "data": {
    "values": [
      {
        "Trial A": -2.9920805103614234,
        "Trial B": -5.007437204443558,
        "Trial C": 2.002018187905394
      },
      {
        "Trial A": -1.731471823260333,
        "Trial B": -4.32763267485067,
        "Trial C": 2.006280602225417
      },
      {
        "Trial A": -1.5603999114082425,
        "Trial B": -4.1063519067301115,
        "Trial C": 2.012536370230117
      },
      {
        "Trial A": -1.4719600496718341,
        "Trial B": -3.892519256743311,
        "Trial C": 2.0246963245480405
      },
      {
        "Trial A": -1.389060325484105,
        "Trial B": -3.7339885232798236,
        "Trial C": 2.0329196002710486
      },
      {
        "Trial A": -1.3320682934384105,
        "Trial B": -3.650411012230142,
        "Trial C": 2.0430671594502043
      },
      {
        "Trial A": -1.258077478203514,
        "Trial B": -3.592216700957609,
        "Trial C": 2.0563545078238485
      },
      {
        "Trial A": -1.2053797755457345,
        "Trial B": -3.525605808882837,
        "Trial C": 2.0675460111452204
      },
      {
        "Trial A": -1.1653911181535381,
        "Trial B": -3.426088444975994,
        "Trial C": 2.074603955736612
      },
      {
        "Trial A": -1.1280678775412292,
        "Trial B": -3.3556426652104716,
        "Trial C": 2.0843364000330635
      },
      {
        "Trial A": -1.0747392492652745,
        "Trial B": -3.3132498054977093,
        "Trial C": 2.0922885749825237
      },
      {
        "Trial A": -1.0584267395139648,
        "Trial B": -3.2576379307260512,
        "Trial C": 2.1051109240003028
      },
      {
        "Trial A": -0.9825527010670889,
        "Trial B": -3.190848938640088,
        "Trial C": 2.112898088682294
      },
      {
        "Trial A": -0.9210517302078746,
        "Trial B": -3.1437358769259216,
        "Trial C": 2.1216111849410755
      },
      {
        "Trial A": -0.8741094127365914,
        "Trial B": -3.087036619245373,
        "Trial C": 2.1347244901101816
      },
      {
        "Trial A": -0.8309377141400818,
        "Trial B": -3.05422940262008,
        "Trial C": 2.1557048239718752
      },
      {
        "Trial A": -0.8086980507424955,
        "Trial B": -2.9775509570250733,
        "Trial C": 2.167214356503413
      },
      {
        "Trial A": -0.7773712117529463,
        "Trial B": -2.9477922964142085,
        "Trial C": 2.1731865950361753
      },
      {
        "Trial A": -0.7391739935957065,
        "Trial B": -2.9236277663441896,
        "Trial C": 2.1834103834281104
      },
      {
        "Trial A": -0.7163784394682,
        "Trial B": -2.852582979055948,
        "Trial C": 2.1910512823546853
      },
      {
        "Trial A": -0.6914921915799892,
        "Trial B": -2.813435337108591,
        "Trial C": 2.1992430330255774
      },
      {
        "Trial A": -0.6690117753209903,
        "Trial B": -2.781135620784436,
        "Trial C": 2.2094879251374877
      },
      {
        "Trial A": -0.6423906125464446,
        "Trial B": -2.7441885031984468,
        "Trial C": 2.217014478151308
      },
      {
        "Trial A": -0.6156480478663496,
        "Trial B": -2.7251759416247907,
        "Trial C": 2.2273327566703442
      },
      {
        "Trial A": -0.5952902810466774,
        "Trial B": -2.700026110222513,
        "Trial C": 2.2389724465239644
      },
      {
        "Trial A": -0.5549619207406208,
        "Trial B": -2.637425202610018,
        "Trial C": 2.2487079265528735
      },
      {
        "Trial A": -0.527034586457933,
        "Trial B": -2.6042170903707182,
        "Trial C": 2.257190310444394
      },
      {
        "Trial A": -0.4984346311397461,
        "Trial B": -2.5646949637556378,
        "Trial C": 2.2717689981360367
      },
      {
        "Trial A": -0.47396414240582097,
        "Trial B": -2.550729013276376,
        "Trial C": 2.2826286064125143
      },
      {
        "Trial A": -0.46708148442884784,
        "Trial B": -2.5202175082410005,
        "Trial C": 2.2911561241486185
      },
      {
        "Trial A": -0.4552828229733322,
        "Trial B": -2.497855888937913,
        "Trial C": 2.302486592782179
      },
      {
        "Trial A": -0.43436410413402526,
        "Trial B": -2.4601196242315813,
        "Trial C": 2.309407406130177
      },
      {
        "Trial A": -0.4031739114786699,
        "Trial B": -2.4224028199936978,
        "Trial C": 2.318224854061052
      },
      {
        "Trial A": -0.376166077989041,
        "Trial B": -2.3905875257556106,
        "Trial C": 2.33229046285205
      },
      {
        "Trial A": -0.3552341181789341,
        "Trial B": -2.3519618625048735,
        "Trial C": 2.3386900417348695
      },
      {
        "Trial A": -0.3406964220143568,
        "Trial B": -2.327969005966306,
        "Trial C": 2.3485273692152173
      },
      {
        "Trial A": -0.32166416479848525,
        "Trial B": -2.3078478234176956,
        "Trial C": 2.3602344243837785
      },
      {
        "Trial A": -0.3064236972873419,
        "Trial B": -2.287335197862673,
        "Trial C": 2.3692433059543765
      },
      {
        "Trial A": -0.285483579986889,
        "Trial B": -2.2707043565409886,
        "Trial C": 2.377283207638597
      },
      {
        "Trial A": -0.2639484967026465,
        "Trial B": -2.2533858221811918,
        "Trial C": 2.3867325558376455
      },
      {
        "Trial A": -0.23529593545965466,
        "Trial B": -2.233356438661789,
        "Trial C": 2.3993127507962773
      },
      {
        "Trial A": -0.20097865701512546,
        "Trial B": -2.215149211335598,
        "Trial C": 2.407710737860619
      },
      {
        "Trial A": -0.1700763899649679,
        "Trial B": -2.1775991456684114,
        "Trial C": 2.4191834033205524
      },
      {
        "Trial A": -0.15586861470185004,
        "Trial B": -2.164726193887353,
        "Trial C": 2.427767215748738
      },
      {
        "Trial A": -0.1372340658772966,
        "Trial B": -2.146533750083727,
        "Trial C": 2.4377062172756343
      },
      {
        "Trial A": -0.11768915181853729,
        "Trial B": -2.1239758859188003,
        "Trial C": 2.449232976640366
      },
      {
        "Trial A": -0.08210940377954545,
        "Trial B": -2.098071625257631,
        "Trial C": 2.4572687448708344
      },
      {
        "Trial A": -0.07095004181722714,
        "Trial B": -2.075196463052107,
        "Trial C": 2.467248128887042
      },
      {
        "Trial A": -0.05111773808249442,
        "Trial B": -2.0458814994311707,
        "Trial C": 2.4802517601342435
      },
      {
        "Trial A": -0.02828313723116236,
        "Trial B": -2.016691147601642,
        "Trial C": 2.4916572791498344
      },
      {
        "Trial A": -0.01939148077408072,
        "Trial B": -1.9900915370129133,
        "Trial C": 2.5016895411909355
      },
      {
        "Trial A": -0.002521412504774076,
        "Trial B": -1.9561640645127119,
        "Trial C": 2.5105272823648335
      },
      {
        "Trial A": 0.015214318882502709,
        "Trial B": -1.9330481909417527,
        "Trial C": 2.520966794635834
      },
      {
        "Trial A": 0.03567322732329613,
        "Trial B": -1.905179802993091,
        "Trial C": 2.528655513369829
      },
      {
        "Trial A": 0.05434326181536148,
        "Trial B": -1.8904850677429073,
        "Trial C": 2.5372896332558135
      },
      {
        "Trial A": 0.0812082963003519,
        "Trial B": -1.85902050191046,
        "Trial C": 2.5446165474480305
      },
      {
        "Trial A": 0.09697763288962838,
        "Trial B": -1.8442870981091213,
        "Trial C": 2.562134642531766
      },
      {
        "Trial A": 0.1131125198386041,
        "Trial B": -1.815251486453477,
        "Trial C": 2.5710323970637203
      },
      {
        "Trial A": 0.12638292237872506,
        "Trial B": -1.7819024684486688,
        "Trial C": 2.580269738112781
      },
      {
        "Trial A": 0.14957937502251734,
        "Trial B": -1.759181309050856,
        "Trial C": 2.5943252100646808
      },
      {
        "Trial A": 0.16709013982957466,
        "Trial B": -1.7281930280973796,
        "Trial C": 2.6013354550424457
      },
      {
        "Trial A": 0.1886090411333894,
        "Trial B": -1.7121282553480626,
        "Trial C": 2.60800094958981
      },
      {
        "Trial A": 0.2151764217005678,
        "Trial B": -1.689749649966775,
        "Trial C": 2.614288924245411
      },
      {
        "Trial A": 0.23877661259146027,
        "Trial B": -1.6491812257243867,
        "Trial C": 2.6239399839901942
      },
      {
        "Trial A": 0.2708075547921658,
        "Trial B": -1.605317368403327,
        "Trial C": 2.6377545093609696
      },
      {
        "Trial A": 0.29596968238274024,
        "Trial B": -1.575072703715482,
        "Trial C": 2.6492033461475
      },
      {
        "Trial A": 0.3164962463851672,
        "Trial B": -1.5470464649397107,
        "Trial C": 2.6598024519300614
      },
      {
        "Trial A": 0.3309513753407172,
        "Trial B": -1.5214256263416992,
        "Trial C": 2.6760244153658728
      },
      {
        "Trial A": 0.3446887972630047,
        "Trial B": -1.487949487629403,
        "Trial C": 2.686124742149673
      },
      {
        "Trial A": 0.3923472803906479,
        "Trial B": -1.4522250098483411,
        "Trial C": 2.6933106747045334
      },
      {
        "Trial A": 0.4212071741463871,
        "Trial B": -1.42764674663949,
        "Trial C": 2.702972085588218
      },
      {
        "Trial A": 0.4437643679413411,
        "Trial B": -1.390635597085511,
        "Trial C": 2.715983143080613
      },
      {
        "Trial A": 0.47699309774060006,
        "Trial B": -1.3753068801245614,
        "Trial C": 2.724099239300397
      },
      {
        "Trial A": 0.49786211709409167,
        "Trial B": -1.3451027219491485,
        "Trial C": 2.7407327489493234
      },
      {
        "Trial A": 0.5209871529359764,
        "Trial B": -1.3148841898483397,
        "Trial C": 2.754691181251933
      },
      {
        "Trial A": 0.5476585953870459,
        "Trial B": -1.28090020524892,
        "Trial C": 2.7637811924923104
      },
      {
        "Trial A": 0.5699319608866455,
        "Trial B": -1.25581904535684,
        "Trial C": 2.7710499054248405
      },
      {
        "Trial A": 0.595071546834937,
        "Trial B": -1.2134557877708336,
        "Trial C": 2.7762970948813104
      },
      {
        "Trial A": 0.6255301129382166,
        "Trial B": -1.1846588004771685,
        "Trial C": 2.785728451585025
      },
      {
        "Trial A": 0.65758817746535,
        "Trial B": -1.1465193841241912,
        "Trial C": 2.7944501793465046
      },
      {
        "Trial A": 0.6903921708249777,
        "Trial B": -1.1056293901809924,
        "Trial C": 2.8025128581584857
      },
      {
        "Trial A": 0.7086230290895082,
        "Trial B": -1.0699672304097476,
        "Trial C": 2.816854823319071
      },
      {
        "Trial A": 0.7325672645438209,
        "Trial B": -1.034729015420874,
        "Trial C": 2.8273754191627667
      },
      {
        "Trial A": 0.778531566900967,
        "Trial B": -0.9873859662107585,
        "Trial C": 2.8376689499084846
      },
      {
        "Trial A": 0.8086515072198301,
        "Trial B": -0.9461811950496086,
        "Trial C": 2.847293138475806
      },
      {
        "Trial A": 0.8431308380736617,
        "Trial B": -0.9243071607380992,
        "Trial C": 2.856660209306999
      },
      {
        "Trial A": 0.8689220340963454,
        "Trial B": -0.8688693438085432,
        "Trial C": 2.8678599789826538
      },
      {
        "Trial A": 0.8996823966799863,
        "Trial B": -0.8338467443886657,
        "Trial C": 2.875992916097936
      },
      {
        "Trial A": 0.9258485479123791,
        "Trial B": -0.7923838222084932,
        "Trial C": 2.88437804251862
      },
      {
        "Trial A": 0.9559979103903102,
        "Trial B": -0.7242194449341406,
        "Trial C": 2.893195971177109
      },
      {
        "Trial A": 1.0068383751101257,
        "Trial B": -0.6688998984797321,
        "Trial C": 2.9031207306919757
      },
      {
        "Trial A": 1.0694127461712446,
        "Trial B": -0.6015833768083174,
        "Trial C": 2.911282428094231
      },
      {
        "Trial A": 1.1326018857779934,
        "Trial B": -0.559746662589798,
        "Trial C": 2.9243687635831366
      },
      {
        "Trial A": 1.2084948576865497,
        "Trial B": -0.4659099739783863,
        "Trial C": 2.9318329285685043
      },
      {
        "Trial A": 1.2527494324702935,
        "Trial B": -0.40666933455475984,
        "Trial C": 2.944218865235456
      },
      {
        "Trial A": 1.3305115841653463,
        "Trial B": -0.351297788196938,
        "Trial C": 2.9544073996926983
      },
      {
        "Trial A": 1.4057454006537355,
        "Trial B": -0.26322645361937447,
        "Trial C": 2.9597605648274032
      },
      {
        "Trial A": 1.4698234132208312,
        "Trial B": -0.100848647328615,
        "Trial C": 2.9675791694972604
      },
      {
        "Trial A": 1.6074072429396256,
        "Trial B": 0.015526019801436093,
        "Trial C": 2.9805393288296216
      },
      {
        "Trial A": 1.9309643128605425,
        "Trial B": 0.3861775776121797,
        "Trial C": 2.9925613567516414
      }
    ]
  },
  "encoding": {
    "x": {
      "field": "Trial A",
      "type": "quantitative"
    },
    "y": {
      "field": "Trial C",
      "type": "quantitative"
    }
  },
  "mark": "point"
}